src/Maestro/Maestro.Common/AppCredential.cs

```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Azure.Core;
+using Azure.Identity;
+
+namespace Microsoft.DotNet.Maestro.Common;
+
+/// <summary>
+/// A credential for authenticating against Azure applications.
+/// </summary>
+public class AppCredential : TokenCredential
+{
+    private static readonly string AUTH_CACHE = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".darc");
+    private static readonly string AUTH_RECORD_PREFIX = ".auth-record";
+
+    private const string TENANT_ID = "72f988bf-86f1-41af-91ab-2d7cd011db47";
+
+    private readonly TokenRequestContext _requestContext;
+    private readonly TokenCredential _tokenCredential;
+
+    private AppCredential(TokenCredential credential, TokenRequestContext requestContext)
+    {
+        _requestContext = requestContext;
+        _tokenCredential = credential;
+    }
+
+    public override AccessToken GetToken(TokenRequestContext _, CancellationToken cancellationToken)
+    {
+        // We hardcode the request context as we know which scopes we need to invoke in each scenario (user vs daemon)
+        return _tokenCredential.GetToken(_requestContext, cancellationToken);
+    }
+
+    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext _, CancellationToken cancellationToken)
+    {
+        // We hardcode the request context as we know which scopes we need to invoke in each scenario (user vs daemon)
+        return _tokenCredential.GetTokenAsync(_requestContext, cancellationToken);
+    }
+
+    /// <summary>
+    /// Use this for user-based flows.
+    /// </summary>
+    public static AppCredential CreateUserCredential(string appId, string userScope = ".default")
+    {
+        var requestContext = new TokenRequestContext(new string[] { $"api://{appId}/{userScope}" });
+
+        string authRecordPath = Path.Combine(AUTH_CACHE, $"{AUTH_RECORD_PREFIX}-{appId}");
+        var credential = GetInteractiveCredential(appId, requestContext, authRecordPath);
+
+        return new AppCredential(credential, requestContext);
+    }
+
+    /// <summary>
+    /// Create interactive credential from an authentication record stored in local cache
+    /// Authentication record is a set of app and user-specific metadata used by the library to authenticate
```

Comment:
I think this should say it gets it through the interactive browser flow but tries to do the cache first.

Reply:
Good point, will clarify